obs-studio-server/source/memory-manager.cpp

```diff
@@ -148,218 +155,171 @@ void MemoryManager::addCachedMemory(source_info *si)
+			si.mtx.lock();
+			calculateRawSize(si); // This also sets 'si.have_video'
 
-			if (si->size || !si->have_video) {
-				si->mtx.unlock();
-				mtx.unlock();
+			if (si.size || !si.have_video) {
+				si.mtx.unlock();
 				break;
 			}
-			si->mtx.unlock();
-			mtx.unlock();
+			si.mtx.unlock();
 
 			retry--;
 			std::this_thread::sleep_for(std::chrono::milliseconds(500));
 		}
 	}
-	std::unique_lock<std::mutex> ulock(mtx);
-	si->mtx.lock();
-	if (!si->size) {
-		si->mtx.unlock();
+
+	std::unique_lock mtx_lock(mtx, std::defer_lock);
+	std::unique_lock si_mtx_lock(si.mtx, std::defer_lock);
+	std::lock(mtx_lock, si_mtx_lock);
+
+	if (!si.size) {
 		return;
 	}
-	bool should_cache = shouldCacheSource(si);
-	si->mtx.unlock();
 
+	const bool should_cache = shouldCacheSource(si);
 	if (should_cache)
 		addCachedMemory(si);
 	else
 		removeCachedMemory(si, true);
 }
 
+// Not thread safe, should be called with locked 'mtx'
 void MemoryManager::updateSettings(obs_source_t *source)
 {
 	auto it = sources.find(obs_source_get_name(source));
-
 	if (it == sources.end())
 		return;
 
-	it->second->workers.push_back(std::thread(&MemoryManager::sourceManager, this, it->second));
+	std::unique_lock ulock(it->second->mtx);
+	it->second->workers.push_back(std::thread(&MemoryManager::sourceManager, this, std::ref(*it->second)));
 }
 
 void MemoryManager::updateSourceCache(obs_source_t *source)
 {
-	std::unique_lock<std::mutex> ulock(mtx);
+	std::unique_lock ulock(mtx);
 
 	updateSettings(source);
 }
 
-void MemoryManager::updateSourcesCache(void)
+void MemoryManager::updateSourcesCache()
 {
-	std::unique_lock<std::mutex> ulock(mtx);
+	std::unique_lock ulock(mtx);
 
-	for (auto data : sources)
+	for (const auto &data : sources)
 		updateSettings(data.second->source);
 }
 
-void MemoryManager::registerSource(obs_source_t *source)
+bool MemoryManager::isSourceValid(obs_source_t *source) const
 {
 	if (!source)
-		return;
+		return false;
 
 	const char *source_id = obs_source_get_id(source);
-	if (!source_id || strcmp(obs_source_get_id(source), "ffmpeg_source"))
+	if (!source_id)
+		return false;
+
+	if (strcmp(source_id, "ffmpeg_source") != 0)
+		return false;
+
+	return true;
+}
+
+void MemoryManager::registerSource(obs_source_t *source)
+{
+	if (!isSourceValid(source)) {
 		return;
+	}
 
-	std::unique_lock<std::mutex> ulock(mtx);
+	std::unique_lock ulock(mtx);
 
 	source_info *si = new source_info;
-	si->cached = false;
-	si->size = 0;
 	si->source = source;
 	sources.emplace(obs_source_get_name(source), si);
 	updateSource(source, false);
-	if (!watcher.running) {
-		watcher.running = true;
-		watcher.worker = std::thread(&MemoryManager::monitorMemory, this);
-	}
 }
 
 void MemoryManager::unregisterSource(obs_source_t *source)
 {
-	if (!source)
-		return;
-	const char *source_id = obs_source_get_id(source);
-	if (!source_id)
-		return;
-	if (strcmp(source_id, "ffmpeg_source") != 0)
+	if (!isSourceValid(source)) {
 		return;
+	}
 
-	mtx.lock();
```

Comment:
[OLD CODE]
Multiple data races here as we do not hold  `mtx` and `it->second.mtx` mutex until deinit ends for usability reasons.
Few the most prominent:
- While someone is called `unregisterSource` it is possible to add new `worker` while we are waiting for threads to join => possibly corrupted `std::vector` of `workers` during iteration
-  Call join from multiple threads is undefined behavior (in reality dead lock)
-  Guaranteed crash if 2 different threads called `unregisterSource` at the same time for the same source and second thread not hanged on previous `join` attempt. It is possible BTW in real life, but chance is very little.